### PR TITLE
Refined documentation for functions, join path bugfix & version bump

### PR DESCRIPTION
**Pull request by akshayrane**
*Issue #, if available:*

*Description of changes:*

1. Updated documentation for functions to have a custom message shown at each condition failure
2. Fixed a bug with `join` function not returning the path in template
3. Bumped up the version to 3.0.0

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by rpdutt — APPROVED** [review]

File: docs/KNOWN_ISSUES.md [hunk 1]
```diff
@@ -35,3 +35,4 @@ let api_gws = Resources.*[ Type == 'AWS::ApiGateway::RestApi' ]
 3. When performing `!=` comparison, if the values are incompatible like comparing a `string` to `int`, an error is thrown internally but currently suppressed and converted to `false` to satisfy the requirements of Rust’s [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html). We are tracking to release a fix for this issue soon.
 4. `exists` and `empty` checks do not display the JSON pointer path inside the document in the error messages. Both these clauses often have retrieval errors which does not maintain this traversal information today. We are tracking to resolve this issue. 
 5. Currently, for `string` literals, Guard does not support embedded escaped strings. We are tracking to resolve this issue soon.
+6. We have support for built-in functions, however, this is currently limited to assignment of the return value to a variable. For example, we can use a function and assign its result to a variable such as `let no_of_instances = count(Instances.*)` and then this variable can be used elsewhere in the conditions such as `%no_of_instances < 2`. However, we **cannot** re-write the same condition as `count(Instances.*) < 2` at this point of time.
```
Comment (joshfried-aws) [inline docs/KNOWN_ISSUES.md:38]: I think we can explain this more thorough. We should probably mention that anytime a function is called it needs to be assigned before it is used anywhere. I.e parameterized rules, other function params, clauses, etc..

**Review by razcloud — APPROVED** [review]

**Review by joshfried-aws — CHANGES_REQUESTED** [review]

File: guard/src/rules/functions/strings.rs [hunk 2]
```diff
@@ -177,13 +177,23 @@ pub(crate) fn join(args: &[QueryResult], delimiter: &str) -> crate::rules::Resul
...
-                    "Joining non unresolved values is not allowed {}, unsatisfied part {}",
+                    "Joining unresolved values is not allowed {}, unsatisfied part {}",
                     ur.traversed_to, ur.remaining_query
                 )));
             }
         }
     }
-    Ok(PathAwareValue::String((Path::root(), aggr)))
+
+    match args.is_empty() {
```
Comment (joshfried-aws) [inline guard/src/rules/functions/strings.rs:187]: Should we add an integration test for this like I did with count yesterday?

Comment (akshayrane) [reply guard/src/rules/functions/strings.rs:187]: good catch, I was planning to, but over the course of raising the PR I forgot to.

File: docs/KNOWN_ISSUES.md (hunk 1 above)
Comment (akshayrane) [reply docs/KNOWN_ISSUES.md]: added explanation

**Review by joshfried-aws — APPROVED** [review]

**Pull request merged**